akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -103,33 +105,79 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
+        val upgradeSettings = req.headers.collect {
+          case raw: RawHeader if raw.lowercaseName == Http2SettingsHeader.name =>
+            Http2SettingsHeader.parse(raw.value)
+        }
+
+        upgradeSettings match {
+          // Must be exactly one
+          case immutable.Seq(Success(http2settings)) =>
+            // TODO remove duplication?
+
+            // TODO inject the HTTP2-Settings -
+            // FIXME: blindly guessed this was the right thing to do...
+            // pre-existing FIXME: parallelism should maybe kept in track with SETTINGS_MAX_CONCURRENT_STREAMS so that we don't need
             // to buffer requests that cannot be handled in parallel
-            .via(Http2Blueprint.handleWithStreamIdHeader(parallelism)(handler)(system.dispatcher))
-            .joinMat(Http2Blueprint.serverStack(settings, log))(Keep.left))
-
-        // TODO consume the HTTP2-Settings header
-        // TODO consume the request entity and make sure it is inserted as an HTTP2 request
-        // https://http2.github.io/http2-spec/#rfc.section.3.2
-
-        Future.successful(
-          HttpResponse(
-            StatusCodes.SwitchingProtocols,
-            immutable.Seq[HttpHeader](
-              ConnectionUpgradeHeader,
-              UpgradeHeader,
-              // TODO add attributes from the request
-              UpgradeToOtherProtocolHeader(serverLayer)
+            val effectiveParallelism =
+              math.min(
+                parallelism,
+                http2settings.find(_.identifier == SETTINGS_MAX_CONCURRENT_STREAMS) match {
+                  case Some(value) => value.value
+                  case None => Int.MaxValue // FIXME can it ever be missing?
+                }
+              )
+
+            val injectedRequest =
+              if (req.method != HttpMethods.OPTIONS)
+                // inject the actual upgrade request with a stream identifier of 1
+                // https://http2.github.io/http2-spec/#rfc.section.3.2
+                Source.single(req.addHeader(Http2StreamIdHeader(1)))
+              else {
+                // No 100% sure about this but from the spec:
+                // If concurrency of an initial request with subsequent requests is important, an OPTIONS request
+                // can be used to perform the upgrade to HTTP/2, at the cost of an additional round trip.
+                req.discardEntityBytes()
+                Source.empty[HttpRequest]
+              }
+
+            val serverLayer: Flow[ByteString, ByteString, Future[Done]] = Flow.fromGraph(
+              Flow[HttpRequest]
+                .watchTermination()(Keep.right)
+                .merge(injectedRequest)
+                .via(Http2Blueprint.handleWithStreamIdHeader(effectiveParallelism)(handler)(system.dispatcher))
+                .joinMat(Http2Blueprint.serverStack(settings, log))(Keep.left))
+
+
+            // TODO do not respond until the potential http2 request entity was consumed?
```

Comment:
I tried this but it deadlocked with nghttp, I'm wondering if the do-not-send-http2 frames part is more about the client than what the server does?